File: IRCClient/src/main/java/net/darkmeow/irc/client/network/handle/encryption/HandleClientEncryptionInbound.java
package net.darkmeow.irc.client.network.handle.encryption;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.util.CharsetUtil;
import net.darkmeow.irc.client.network.IRCClientConnection;
import net.darkmeow.irc.utils.EncryptUtils;

public final class HandleClientEncryptionInbound extends ChannelInboundHandlerAdapter {

    public final IRCClientConnection client;

    public HandleClientEncryptionInbound(IRCClientConnection client) {
        this.client = client;
    }

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object data) throws Exception {
        if (data instanceof ByteBuf) {
            final ByteBuf buf = (ByteBuf) data;
            try {
                super.channelRead(ctx, EncryptUtils.decrypt(buf.toString(CharsetUtil.UTF_8), client.key));
            } finally {
                buf.release();
            }
        }
    }

}

File: IRCServer/src/main/kotlin/net/darkmeow/irc/network/handles/encryption/HandleClientEncryptionInbound.kt
package net.darkmeow.irc.network.handles.encryption

import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.util.CharsetUtil
import net.darkmeow.irc.network.NetworkManager
import net.darkmeow.irc.utils.EncryptUtils
import javax.crypto.BadPaddingException

class HandleClientEncryptionInbound(private val manager: NetworkManager): ChannelInboundHandlerAdapter() {

    override fun channelRead(ctx: ChannelHandlerContext, data: Any) {
        if (data is ByteBuf) {
            runCatching {
                super.channelRead(
                    ctx,
                    EncryptUtils.decrypt(data.toString(CharsetUtil.UTF_8), manager.base.configManager.configs.ircServer.key)
                )
            }
                .onFailure { e ->
                    if (e !is BadPaddingException && e !is IllegalArgumentException) {
                        throw e
                    }
                }
                .also {
                    data.release()
                }
        }
    }

}